### PR TITLE
Fix a crash related to invalid number literals.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1432,11 +1434,17 @@ void ExpressionCompiler::appendExternalFunctionCall(
...
-	bool manualFunctionId =
+	bool manualFunctionId = false;
+	if (
 		(funKind == FunctionKind::Bare || funKind == FunctionKind::BareCallCode || funKind == FunctionKind::BareDelegateCall) &&
-		!_arguments.empty() &&
-		_arguments.front()->annotation().type->mobileType()->calldataEncodedSize(false) ==
+		!_arguments.empty()
+	)
+	{
+		solAssert(_arguments.front()->annotation().type->mobileType(), "");
```
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp:1448]: The combination looks strange
- the code allows empty arguments; but
- the code does not allow an existing first argument to be not-mobile.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -986,10 +990,16 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
...
+				if (_tuple.isInlineArray())
+				{
+					if ((i == 0 || inlineArrayType) && !types[i]->mobileType())
+						fatalTypeError(components[i]->location(), "Invalid mobile type.");
+
+					if (i == 0)
+						inlineArrayType = types[i]->mobileType();
+					else if (inlineArrayType)
+						inlineArrayType = Type::commonType(inlineArrayType, types[i]->mobileType());
+				}
```
Comment (pirapira) [inline libsolidity/analysis/TypeChecker.cpp:1002]: 18 lines below, there might be a segmentation fault when components.size() == 0.  The access to `components[1]` is not guarded against `components.size() == 0`.

Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:1002]: This was my brain fart.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -986,10 +990,16 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
...
 				if (_tuple.isInlineArray())
 					solAssert(!!types[i], "Inline array cannot have empty components");
-				if (i == 0 && _tuple.isInlineArray())
-					inlineArrayType = types[i]->mobileType();
-				else if (_tuple.isInlineArray() && inlineArrayType)
-					inlineArrayType = Type::commonType(inlineArrayType, types[i]->mobileType());
+				if (_tuple.isInlineArray())
+				{
+					if ((i == 0 || inlineArrayType) && !types[i]->mobileType())
+						fatalTypeError(components[i]->location(), "Invalid mobile type.");
```
Comment (pirapira) [inline libsolidity/analysis/TypeChecker.cpp:996]: Is this check necessary?  `Type::commonType` returns the null type pointer if one of the arguments is nullptr.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -198,7 +198,9 @@ TypePointer Type::forLiteral(Literal const& _literal)
 
 TypePointer Type::commonType(TypePointer const& _a, TypePointer const& _b)
 {
-	if (_b->isImplicitlyConvertibleTo(*_a))
+	if (!_a || !_b)
+		return TypePointer();
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp:202]: Shall I understand `TypePointer()` as the type of every expression or the type of no expressions?  If there are no values that belong to `TypePointer()`, `TypePointer()` should be implicitly convertible to any type.  If all values belong to `TypePointer()`, every type should be implicitly convertible to `TypePointer()`.

Comment (pirapira) [reply libsolidity/ast/Types.cpp:202]: ... maybe they are just saying "no types".

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:996]: This does produce a better error message.  OK.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 5]
```diff
@@ -1432,11 +1434,17 @@ void ExpressionCompiler::appendExternalFunctionCall(
...
+		manualFunctionId =
+			_arguments.front()->annotation().type->mobileType()->calldataEncodedSize(false) ==
 			CompilerUtils::dataStartOffset;
```
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp:1451]: What happens if the user did not want to specify a function ID manually, but she just wanted the first argument to be 4 bytes long?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1451]: Indeed, this is not possible. In the long run, we should deprecate automatic encoding by `call` and only allow a `bytes` argument.

File: libsolidity/ast/Types.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:202]: `TypePointer()` is an invalid type and here it just means "there is no common type". A better way would be to report an actual error that can be displayed to the user.

**Comment by pirapira** [comment]
I'm wary of the manual function ID detection by looking at the type (it would surprise me to see `bytes4` behave very differently from `bytes5`).  However that's not this PR's fault.

**Review by pirapira — APPROVED** [review]
Though I asked many questions, this PR does not introduce new problems but solve one.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 5 above)
Comment (pirapira) [reply libsolidity/codegen/ExpressionCompiler.cpp:1451]: I didn't quite understand the last sentence (automatic encoding... of `bytes` into `bytes5`, or automatic encoding of funciton IDs, or something else??).  Will you open an issue about this?

**Pull request merged**